rosbag2/src/rosbag2/types/ros2_message.cpp

```diff
@@ -0,0 +1,267 @@
+        auto data_vector = static_cast<std::vector<uint16_t> *>(data);
+        std::vector<uint16_t> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_INT32: {
+        auto data_vector = static_cast<std::vector<int32_t> *>(data);
+        std::vector<int32_t> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT32: {
+        auto data_vector = static_cast<std::vector<uint32_t> *>(data);
+        std::vector<uint32_t> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_INT64: {
+        auto data_vector = static_cast<std::vector<int64_t> *>(data);
+        std::vector<int64_t> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT64: {
+        auto data_vector = static_cast<std::vector<uint64_t> *>(data);
+        std::vector<uint64_t> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING: {
+        auto data_vector = static_cast<std::vector<std::string> *>(data);
+        std::vector<std::string> empty;
+        data_vector->swap(empty);
+        break;
+      }
+    case rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE: {
+        auto data_vector = static_cast<std::vector<uint8_t> *>(data);
+
+        auto nested_ts =
+          static_cast<const rosidl_typesupport_introspection_cpp::MessageMembers *>(
+          member.members_->data);
+        auto size = member.size_function(data);
+
+        for (size_t j = 0; j < size; ++j) {
+          auto nested_member = data_vector->data() + j * nested_ts->size_of_;
+          deallocate_ros2_message_part(nested_member, nested_ts);
+        }
+
+        data_vector->resize(0);
+        data_vector->shrink_to_fit();
+        delete data_vector->data();
+        break;
+      }
+  }
+}
+
+void allocate_internal_types(
+  void * msg, const rosidl_typesupport_introspection_cpp::MessageMembers * members)
+{
+  for (size_t i = 0; i < members->member_count_; ++i) {
+    auto member = members->members_[i];
+    void * message_member = static_cast<uint8_t *>(msg) + member.offset_;
+
+    if ((member.is_array_ && member.array_size_ == 0) || member.is_upper_bound_) {
+      allocate_vector(message_member, member);
+    } else if (member.is_array_ && member.array_size_ > 0) {
+      allocate_array(message_member, member);
+    } else {
+      allocate_element(message_member, member);
+    }
+  }
+}
+
+void allocate_vector(void * data, rosidl_typesupport_introspection_cpp::MessageMember member)
+{
+  // This is necessary because initialization otherwise fails for MSVC++ compiled builds
+  if (member.type_id_ == rosidl_typesupport_introspection_cpp::ROS_TYPE_BOOL) {
+    new (data) std::vector<bool>();
+  }
```

Comment:
I have a very bad feeling here and this sure isn't going to fly. We need to actually call placement new for `std::vector<T>` for all types that we need. Anything else will break on some system at some point in time in a very odd way, because there is no guarantee at all that just using zero-allocated memory can be reinterpreted as an empty vector. This might even break already when activating more optimisation in the compiler. 

This means we have a real problem for all the non-primitive types. There seems to be no way to reliably allocate a nested message given only the introspection typesupport.

Reply:
I found at least two places within ROS 2 where we actually do exactly that with our messages:

The first one is used in fastrtps_dynamic when parsing messages from CDR:
https://github.com/ros2/rmw_fastrtps/blob/master/rmw_fastrtps_dynamic_cpp/include/rmw_fastrtps_dynamic_cpp/TypeSupport_impl.hpp#L271-L287

We also do it in the message generation step ourselves:
https://github.com/ros2/rosidl/blob/master/rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.em#L43-L53

So I think this goes okay here as long as the message generation of rosidl_typesupport_cpp does the same thing.

Reply:
We changed this to always create a vector using placement new.